src/http.cpp

```diff
@@ -773,51 +767,74 @@ m_off_t HttpReqUL::transferred(MegaClient* client)
 
 SpeedController::SpeedController()
 {
-    partialBytes = 0;
-    meanSpeed = 0;
-    lastUpdate = 0;
-    speedCounter = 0;
+    memset(circularBuf.data(), 0, sizeof(circularBuf));
+}
+
+void SpeedController::requestStarted()
+{
+    requestPos = 0;
+    requestStart = lastRequestUpdate = Waiter::ds;
+}
+
+m_off_t SpeedController::requestProgressed(m_off_t newPos)
+{
+    if (newPos > requestPos)
+    {
+        m_off_t delta = newPos - requestPos;
+        calculateSpeed(delta);
+        requestPos = newPos;
+        lastRequestUpdate = Waiter::ds;
+        return delta;
+    }
+    return 0;
+}
+
+m_off_t SpeedController::lastRequestSpeed()
+{
+    dstime deltaDs = lastRequestUpdate - requestStart;
+    return requestPos * 10 / (deltaDs ? deltaDs : 1);
+}
+
+dstime SpeedController::requestElapsedDs()
+{
+    return Waiter::ds - requestStart;
 }
 
 m_off_t SpeedController::calculateSpeed(long long numBytes)
 {
     dstime currentTime = Waiter::ds;
-    if (numBytes <= 0 && lastUpdate == currentTime)
+    if (numBytes <= 0 && lastCalcTime == currentTime)
     {
-        return (partialBytes * 10) / SPEED_MEAN_INTERVAL_DS;
+        return (circularCurrentSum * 10) / SPEED_MEAN_MAX_INTERVAL_DS;
     }
 
-    while (transferBytes.size())
+    for (int i = SPEED_MEAN_MAX_INTERVAL_DS; i--; )
     {
-        map<dstime, m_off_t>::iterator it = transferBytes.begin();
-        dstime deltaTime = currentTime - it->first;
-        if (deltaTime < SPEED_MEAN_INTERVAL_DS)
+        if (circularCurrentTime < currentTime)
         {
-            break;
+            ++circularCurrentTime;
```

Comment:
In case it's been more than one decisecond since the last time we got a call to update the buffer - so we loop up to the size of the buffer max, or until the bucket we've iterated to corresponds to the current decisecond.

Reply:
I see. Thanks for the explanation. Makes perfect sense :+1: